src/Functions/UTCTimestampTransform.cpp

```diff
@@ -0,0 +1,169 @@
+            if (dynamic_cast<const TimezoneMixin *>(arguments[0].get())->hasExplicitTimeZone())
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 1st argument should not have explicit time zone.", name);
+            if (!which_type_second.isString())
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 2nd argument type must be string.", name);
+            DataTypePtr date_time_type;
+            if (which_type_first.isDateTime())
+                date_time_type = std::make_shared<DataTypeDateTime>();
+            else
+            {
+                const DataTypeDateTime64 * date_time_64 = static_cast<const DataTypeDateTime64 *>(arguments[0].get());
+                date_time_type = std::make_shared<DataTypeDateTime64>(date_time_64->getScale());
+            }
+            return date_time_type;
+        }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t) const override
+        {
+            if (arguments.size() != 2)
+                throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {}'s arguments number must be 2.", name);
+            ColumnWithTypeAndName arg1 = arguments[0];
+            ColumnWithTypeAndName arg2 = arguments[1];
+            const auto *  time_zone_col = checkAndGetColumn<ColumnString>(arg2.column.get());
+            if (!time_zone_col)
+            {
+                const auto * time_zone_const_col = checkAndGetColumn<ColumnConst>(arg2.column.get());
+                if (!time_zone_const_col)
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 2nd argument is not string type or const string type.", name);
+                time_zone_col = checkAndGetColumn<ColumnString>(time_zone_const_col->getDataColumn());
+                if (!time_zone_col)
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 2nd argument must be string type or const string type.", name);
+            }
+            auto column = result_type->createColumn();
+            if (WhichDataType(arg1.type).isDateTime())
+            {
+                const auto * date_time_col = checkAndGetColumn<ColumnDateTime>(arg1.column.get());
+                if (!date_time_col)
+                {
+                    const auto * date_time_const_col = checkAndGetColumn<ColumnConst>(arg1.column.get());
+                    if (!date_time_const_col)
+                        throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 1st argument is not datatime type or const type.", name);
+                    date_time_col = checkAndGetColumn<ColumnDateTime>(date_time_const_col->getDataColumn());
+                    if (!date_time_col)
+                        throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Function {}'s 1st argument must be datetime type", name);
+                }
```

Comment:
Same problem with constant here. Underlying column will have size 1.
```
avogar-dev :) select toUTCTimestamp(toDateTime('2020-01-01'), materialize('Asia/Shanghai'))

0 rows in set. Elapsed: 0.001 sec.

Received exception:
Code: 1001. DB::Exception: std::length_error: basic_string. (STD_EXCEPTION)
```

If we want to allow only constant time zone, than we will never get constant first argument and we can avoid this code (because if both arguments will be constant, default implementation for constants will be applied and first argument will be converted to full column with size 1)

Reply:
If we want to allow non-constant time zone, then we should handle and test all 3 cases:
1) arg[0] is not const, arg[1] is not const
2) arg[0] is not const, arg[1] is const
3) arg[0] is const, arg[1] is not const

Case when both arguments are cons will be handled by default implementation for constants and we will get 1) case with 1-size columns